test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/LocalClusterFactory.java

```diff
@@ -0,0 +1,485 @@
+    private static final String TESTS_CLUSTER_PLUGINS_PATH_SYSPROP = "tests.cluster.plugins.path";
+
+    private final Path baseWorkingDir;
+    private final DistributionResolver distributionResolver;
+
+    public LocalClusterFactory(Path baseWorkingDir, DistributionResolver distributionResolver) {
+        this.baseWorkingDir = baseWorkingDir;
+        this.distributionResolver = distributionResolver;
+    }
+
+    @Override
+    public LocalClusterHandle create(LocalClusterSpec spec) {
+        return new LocalClusterHandle(spec.getName(), spec.getNodes().stream().map(Node::new).toList());
+    }
+
+    public class Node {
+        private final LocalNodeSpec spec;
+        private final Path workingDir;
+        private final Path distributionDir;
+        private final Path snapshotsDir;
+        private final Path dataDir;
+        private final Path logsDir;
+        private final Path configDir;
+        private final Path tempDir;
+
+        private boolean initialized = false;
+        private Process process = null;
+        private DistributionDescriptor distributionDescriptor;
+
+        public Node(LocalNodeSpec spec) {
+            this.spec = spec;
+            this.workingDir = baseWorkingDir.resolve(spec.getCluster().getName()).resolve(spec.getName());
+            this.distributionDir = workingDir.resolve("distro"); // location of es distribution files, typically hard-linked
+            this.snapshotsDir = workingDir.resolve("repo");
+            this.dataDir = workingDir.resolve("data");
+            this.logsDir = workingDir.resolve("logs");
+            this.configDir = workingDir.resolve("config");
+            this.tempDir = workingDir.resolve("tmp"); // elasticsearch temporary directory
+        }
+
+        public synchronized void start() {
+            LOGGER.info("Starting Elasticsearch node '{}'", spec.getName());
+
+            if (initialized == false) {
+                LOGGER.info("Creating installation for node '{}' in {}", spec.getName(), workingDir);
+                distributionDescriptor = resolveDistribution();
+                LOGGER.info("Distribution for node '{}': {}", spec.getName(), distributionDescriptor);
+                initializeWorkingDirectory();
+                writeConfiguration();
+                createKeystore();
+                configureSecurity();
+                installPlugins();
+                if (spec.getDistributionType() == DistributionType.INTEG_TEST) {
+                    installModules();
+                }
+                initialized = true;
+            }
+
+            startElasticsearch();
+        }
+
+        public synchronized void stop(boolean forcibly) {
+            if (process != null) {
+                ProcessUtils.stopHandle(process.toHandle(), forcibly);
+                ProcessReaper.instance().unregister(getServiceName());
+            }
+        }
+
+        public void waitForExit() {
```

Comment:
Doesn't this need to be synchronized since it is checking `process`?

Reply:
I don't believe so given it doesn't actually mutate the process.